### ISSUE TITLE
How to optionally reference another package's variable

### ISSUE THREAD

**Issue by DamienCassou** [purcell/package-lint#205]
I feel stupid asking that as I should know the answer already but I don't (or I forgot).

I'm writing a package which makes an /optional/ reference to a variable from another, non Emacs-core, package. I protected the use of the variable within a bound-and-true-p but the byte compiler warns me that the variable is free. I would like to pacify it and tell it it's fine. Do you know how I can declare the external variable without having package-lint complain that the name of the variable should start with the current package name?

**Comment by akirak** [comment]
Hello,

In my experiences, adding a `defvar` form without the default value would suffice:

```emacs-lisp
(defvar package-variable)
```

It must be declared before the function using the variable.

**Comment by purcell** [comment]
Yes, that should indeed pass `package-lint`'s checks. Ideally such forward declarations are extremely rare: better where possible to have more granular packages with explicit dependencies and proper use of `require`.

**Comment by akirak** [comment]
> package-lint complain that the name of the variable should start with the current package name

AFAIK package-lint doesn't complain about variable names. It's a restriction about function names.

**Comment by purcell** [comment]
> AFAIK package-lint doesn't complain about variable names. It's a restriction about function names.

Interesting. That seems like an oversight.

**Comment by purcell** [comment]
I added tests for that case and it looks like everything is already working as I would have wanted: see 30258f5

**Comment by akirak** [comment]
So, does it throw an error if and only if the variable has the default value?

**Comment by purcell** [comment]
> So, does it throw an error about the prefix if and only if the variable has an initial value?

Yep.

**Comment by purcell** [comment]
Oops, looks like I'm mistaken and I've pushed a failing test. I've got a cold so my brain is not at 100%.

**Comment by purcell** [comment]
I've commented that one out for now.

**Comment by akirak** [comment]
@purcell Got it, thanks.

**Comment by DamienCassou** [comment]
> According to my experiences, adding a defvar form without the default value would suffice

thank you!

> Ideally such forward declarations are extremely rare: better where possible to have more granular packages with explicit dependencies and proper use of require.

I completely agree. I think that the [beginend package](https://github.com/DamienCassou/beginend) is an exception to this rule though: it improves navigation of many modes and requires just a few lines for each supported one.

I worked around the package-lint issue by [adding a function to the external package](https://github.com/NicolasPetton/nroam/pull/24) through a PR. This is nicer than accessing private variables and functions anyway.
